### ISSUE TITLE
Brahms needs a sensible build process

### ISSUE THREAD

**Issue by sebjameswml** [sebjameswml/brahms#2]
This will involve reworking the Makescripts using cmake or something similar. Cmake is good candidate as it's well supported both on Windows, Mac and Linux.

**Comment by sebjameswml** [comment]
There are numerous patches for the brahms build which I developed for SpineML/debian_packaging. These are useful, so here are some of them:

add-build-install-path.patch:
```patch
Description: Adds a build install path
 Brahms uses a variable SYSTEMML_INSTALL_PATH within the code to find
 libraries which it dynamically links at runtime. In the upstream
 code, this path is used as the install path. In the Debian package
 build process, it's important to install in a separate location so that
 files can then be copied into the location defined by SYSTEMML_INSTALL_PATH.
 This patch adds SYSTEMML_BUILD_INSTALL_PATH which does just that.
Forwarded: No need to forward, packaging fix only.
Author: Seb James <seb.james@sheffield.ac.uk>
Last-Update: 2014-02-21
--- a/makefile.brahms
+++ b/makefile.brahms
@@ -137,7 +137,17 @@
 ifeq ($(SYSTEMML_INSTALL_PATH),)
 ${error set SYSTEMML_INSTALL_PATH}
 endif
-PATH_SML=${subst $(SPACE),\$(SPACE),${subst \,/,$(SYSTEMML_INSTALL_PATH)}}
+
+# The "build install path" is a location into which the build will be
+# installed, prior to being installed on the system at
+# SYSTEMML_INSTALL_PATH. This is necessary for the Debian packaing
+# system, so that the debuild process can build, then install.  If
+# this variable is unset, set it to SYSTEMML_INSTALL_PATH.
+ifeq ($(SYSTEMML_BUILD_INSTALL_PATH),)
+SYSTEMML_BUILD_INSTALL_PATH=$(SYSTEMML_INSTALL_PATH
+endif
+
+PATH_SML=${subst $(SPACE),\$(SPACE),${subst \,/,$(SYSTEMML_BUILD_INSTALL_PATH)}}
 PATH_BRM=$(PATH_SML)/BRAHMS
 
 # temp path
```
Description: More -w flags for CFLAGS/CXXFLAGS
Author: Seb James <seb.james@sheffield.ac.uk>
Forwarded: No need to forward.
Last-Update: 2014-02-23

add-additional-w-flags-to-cflags.patch
```patch
--- a/makefile.brahms
+++ b/makefile.brahms
@@ -284,7 +284,7 @@
 ifeq ($(EASY_BUILD), 1)
 	CFLAGS=-fPIC -ffast-math $(FLAG_PTHREAD) $(FLAG_ARCH) -Wfatal-errors -fno-strict-aliasing
 else
-	CFLAGS=-fPIC -Werror -Wall -ffast-math $(FLAG_PTHREAD) $(FLAG_ARCH) -Wfatal-errors -fno-strict-aliasing -fstack-protector --param=ssp-buffer-size=4
+	CFLAGS=-fPIC -g -Wformat -Werror=format-security -Werror -Wall -ffast-math $(FLAG_PTHREAD) $(FLAG_ARCH) -Wfatal-errors -fno-strict-aliasing -fstack-protector --param=ssp-buffer-size=4
 	LDFLAGS=-Wl,-z,relro
 endif
 ```

allow-empty-systemml-install-path.patch
```patch
Description: Allow SYSTEMML_INSTALL_PATH to be empty
 Simply prevent brahms from showing an error if this path is empty. This is
 Debian package specific.
Author: Seb James <seb.james@sheffield.ac.uk>
Forwarded: No need to forward.
Last-Update: 2014-02-22

--- a/framework/engine/support/environment.cpp
+++ b/framework/engine/support/environment.cpp
@@ -312,7 +312,7 @@
 
 			//	construct list of namespace roots
 			string NamespaceRoots = brahms::os::getenv("SYSTEMML_INSTALL_PATH");
-			if (!NamespaceRoots.length()) ferr << E_INSTALLATION << "env var SYSTEMML_INSTALL_PATH not set";
+			//if (!NamespaceRoots.length()) ferr << E_INSTALLATION << "env var SYSTEMML_INSTALL_PATH not set";
 			NamespaceRoots += brahms::os::PATH_SEPARATOR;
 			NamespaceRoots += "Namespace";
 			string s = gets("NamespaceRoots");
```

XtAppAddTimeOut patch Gauge.c
```patch
Description: Fix compiler warning in Gauge.c
 A fix to ensure brahms compiles with gcc-4.7
Author: Seb James <seb.james@sheffield.ac.uk>
Forwarded: By email to Ben Mitchinson
Last-Update: 2014-02-21
--- a/framework/execute/gui/Gauge.c
+++ b/framework/execute/gui/Gauge.c
@@ -1084,8 +1084,9 @@
 {
 	gw->gauge.intervalId =
 	  XtAppAddTimeOut(XtWidgetToApplicationContext((Widget)gw),
-	  	gw->gauge.update * MS_PER_SEC, GaugeGetValue,
-		(XtPointer)gw) ;
+	                  gw->gauge.update * MS_PER_SEC,
+	                  (XtTimerCallbackProc)GaugeGetValue,
+	                  (XtPointer)gw) ;
 }
 
 static	void
```

Stuff to do with Namespace:
```patch
Description: Add Debian-specific default path to Namespace
 Namespace is a directory tree shipped as an example for brahms users
 to copy and use. For Debian, I've installed it in /var/lib/brahms.
Author: Seb James <seb.james@sheffield.ac.uk>
Forwarded: No need to forward.
Last-Update: 2014-02-24

--- a/framework/engine/support/environment.cpp
+++ b/framework/engine/support/environment.cpp
@@ -314,7 +314,13 @@
 			string NamespaceRoots = brahms::os::getenv("SYSTEMML_INSTALL_PATH");
 			//if (!NamespaceRoots.length()) ferr << E_INSTALLATION << "env var SYSTEMML_INSTALL_PATH not set";
 			NamespaceRoots += brahms::os::PATH_SEPARATOR;
-			NamespaceRoots += "Namespace";
+                        NamespaceRoots += "var";
+			NamespaceRoots += brahms::os::PATH_SEPARATOR;
+                        NamespaceRoots += "lib";
+			NamespaceRoots += brahms::os::PATH_SEPARATOR;
+                        NamespaceRoots += "brahms";
+			NamespaceRoots += brahms::os::PATH_SEPARATOR;
+                        NamespaceRoots += "Namespace";
 			string s = gets("NamespaceRoots");
 			if (s.length())
 			{
```

add-build-install-path.patch
```patch
Description: Adds a build install path
 Brahms uses a variable SYSTEMML_INSTALL_PATH within the code to find
 libraries which it dynamically links at runtime. In the upstream
 code, this path is used as the install path. In the Debian package
 build process, it's important to install in a separate location so that
 files can then be copied into the location defined by SYSTEMML_INSTALL_PATH.
 This patch adds SYSTEMML_BUILD_INSTALL_PATH which does just that.
Forwarded: No need to forward, packaging fix only.
Author: Seb James <seb.james@sheffield.ac.uk>
Last-Update: 2014-02-21
--- a/makefile.brahms
+++ b/makefile.brahms
@@ -137,7 +137,17 @@
 ifeq ($(SYSTEMML_INSTALL_PATH),)
 ${error set SYSTEMML_INSTALL_PATH}
 endif
-PATH_SML=${subst $(SPACE),\$(SPACE),${subst \,/,$(SYSTEMML_INSTALL_PATH)}}
+
+# The "build install path" is a location into which the build will be
+# installed, prior to being installed on the system at
+# SYSTEMML_INSTALL_PATH. This is necessary for the Debian packaing
+# system, so that the debuild process can build, then install.  If
+# this variable is unset, set it to SYSTEMML_INSTALL_PATH.
+ifeq ($(SYSTEMML_BUILD_INSTALL_PATH),)
+SYSTEMML_BUILD_INSTALL_PATH=$(SYSTEMML_INSTALL_PATH
+endif
+
+PATH_SML=${subst $(SPACE),\$(SPACE),${subst \,/,$(SYSTEMML_BUILD_INSTALL_PATH)}}
 PATH_BRM=$(PATH_SML)/BRAHMS
 
 # temp path
```

allow-arch-bits-set-in-makefile.patch
```patch
--- a/makefile.brahms
+++ b/makefile.brahms
@@ -68,7 +68,11 @@
 ${error Unable to identify platform - modify makefile.brahms}
 endif

+# Has the user defined ARCH_BITS?
+ifeq ($(ARCH_BITS),)
+
 # 64-bit detection (plus define ARCH_NIX)
+# no user-specified value for ARCH_BITS, work out now.
 ifeq ($(ARCH_OS), win)
 POSS64=${shell echo $(PROCESSOR_ARCHITEW6432)$(PROCESSOR_IDENTIFIER) | grep 64}
 else
@@ -93,6 +97,7 @@
 ARCH_BITS=$(TRUE_ARCH_BITS)
 endif
 
+endif # was ARCH_BITS specified by the user?
 
 # collate
 ARCH=$(ARCH_OS)$(ARCH_BITS)
```

bugfix-par-cmd-line-option-with-dash.patch
```patch
--- a/framework/execute/main.cpp
+++ b/framework/execute/main.cpp
@@ -218,8 +218,13 @@
 		if (arg.length() >= 6 && arg.substr(0,6) == "--par-")
 		{
 			string xy = arg.substr(6);
-			size_t pos = xy.find('-');
-			if (pos == string::npos) pos = xy.find('=');
+                        size_t pos = xy.find("MaxThreadCount-");
+                        if (pos != string::npos) { // Then this is the special case --par-MaxThreadCount-N
+                            pos += 14; // length of MaxThreadCount
+                        } else {
+                            // else normal case; no MaxThreadCount found; option is --par-SomeParam=SomeValue
+                            pos = xy.find('=');
+                        }
 			if (pos == string::npos) client_err << "E_INVOCATION: unrecognised option \"" + arg + "\"";
 			instance.execPars += xy.substr(0, pos) + "=" + xy.substr(pos + 1) + ";";
 			createEngine.executionParameters = instance.execPars.c_str();
```

To do with search paths, I think this is something which will be important
modify-search-paths.patch
```patch
Description: Modify paths for components of brahms
 In order to be more Debian friendly.
Author: Seb James <seb.james@sheffield.ac.uk>
Forwarded: No need to forward
Last-Update: 2014-02-21
--- a/framework/engine/support/os.h
+++ b/framework/engine/support/os.h
@@ -71,10 +71,10 @@
 		const char ENV_SEPARATOR[] = ":";
 		const char COMPONENT_DLL_EXT[] = "so";
 		const char COMPONENT_MODULE_FORM_OLD[] = "libbrahms-((CLASS_US)).so.0";
-		const char M_BINDING_MODULE_FORM[] = "/BRAHMS/bindings/component/1258/libbrahms-1258.so.0";
-		const char PY_BINDING_MODULE_FORM[] = "/BRAHMS/bindings/component/1262/libbrahms-1262.so.0";
-		const char CHANNEL_MODULE_FORM[] = "/BRAHMS/bin/libbrahms-channel-((PROTOCOL)).so";
-		const char COMPRESS_MODULE_FORM[] = "/BRAHMS/bin/libbrahms-compress.so";
+		const char M_BINDING_MODULE_FORM[] = "/usr/lib/brahms/bindings/component/1258/libbrahms-1258.so.0";
+		const char PY_BINDING_MODULE_FORM[] = "/usr/lib/brahms/bindings/component/1262/libbrahms-1262.so.0";
+		const char CHANNEL_MODULE_FORM[] = "/usr/lib/brahms/libbrahms-channel-((PROTOCOL)).so";
+		const char COMPRESS_MODULE_FORM[] = "/usr/lib/brahms/libbrahms-compress.so";
 
 	#endif
 
--- a/framework/engine/main/engine.cpp
+++ b/framework/engine/main/engine.cpp
@@ -178,7 +178,7 @@
 
 				//	load install/machine/user-level environment
 				{ FOUT_SECTION("Install-Level")
-					engineData.environment.load(brahms::os::getenv("SYSTEMML_INSTALL_PATH") + "/BRAHMS/brahms.xml", true, fout);
+					engineData.environment.load(brahms::os::getenv("SYSTEMML_INSTALL_PATH") + "/usr/share/brahms/brahms.xml", true, fout);
 				}
 				{ FOUT_SECTION("Machine-Level")
 					engineData.environment.load(brahms::os::getspecialfolder("appdata.machine") + "/brahms.xml", false, fout);
```

**Comment by sebjameswml** [comment]
Commits up to b22a096 implement a cmake based build system which I have used to compile and install brahms and run a SpineML model (using SpineML_2_BRAHMS).
